added parameter in count function

Anukriti12 opened the pull request:
First PR: try_depth yet to be removed

trefis committed:
matching: add a test for do_for_multiple_match flattening failure

Note: this is due to mk_alpha_env raising Cannot_flatten during
splitting/precompilation.

dra27 committed:
Eliminate MKEXE_ANSI from build system

dra27 committed:
Share generation code between ocamltest and utils

ocamltest/ocamltest_config.ml now generated using same make macros as
utils/config.ml.

trefis committed:
matching: argo => arg_id

trefis committed:
matching: try => match with exception

This makes it clearer where the exception comes from.

trefis committed:
matching: push simple types to flatten_pattern

Unfortunately since the function is exposed and used in translcore we
need to keep the generic one, and introduce a flatten_simple_pattern.

trefis committed:
matching: flatten_simple_pattern does not raise

gasche committed:
[minor] printlambda: print the 'tailcall' attribute in the same style as others

gasche committed:
[refactoring] gives tailcall attributes a more standard structure

We want to start allowing more information in the payload of
[@tailcall] attributes (currently no payload is supported), for
example we could consider using [@tailcall false] to ask the code
generator to disable a tail call.

A first required step in this direction is to use a custom datatype to
represent the tail-call attribute, instead of a boolean. This is
consistent with the other application-site
attributes (inline_attribute, specialise_attribute, local_attribute),
so it makes the code more regular -- but the change itself is
boilerplate-y.

gasche committed:
Parmatch.exhaust: single-row optimization

(See testsuite and code comments for an explanation.)

stedolan committed:
More partial application warnings

gasche committed:
Arg: a new Rest_all spec, similar to Rest, that passes all arguments at once

In particular this allow to detect a Rest_all keyword followed by no
arguments at all (an empty list).

garrigue committed:
add explanation for variance flags

xavierleroy committed:
FFI documentation: naked pointers are obsolete

Also: better explain how to encapsulate naked pointers in Abstract blocks.

garrigue committed:
fix number and add examples for #6744

Octachron committed:
Format: margin documentation precision

Octachron committed:
format: comment max indent induced line splits

Octachron committed:
review and change entry for #9618

Octachron committed:
format: document structural box quirk

xavierleroy committed:
Give preallocated atoms the GC color Black

Atoms are zero-sized blocks allocated outside the heap.
It simplifies the GC in no-naked-pointers mode if their headers
have GC color Black, meaning "don't traverse".
In ocamlopt, Black is already used as the color for constant blocks
statically allocated outside the heap.

xavierleroy committed:
No need to special-case zero-sized blocks in no-naked-pointers mode

Now that atoms have black headers, all zero-sized blocks (atoms or
ocamlopt-generated static data) have black headers and will not
be traversed or changed by the major GC.

xavierleroy committed:
New representation of closures, bytecode part

Add a "closure information" field after each code pointer in
a closure.  This field generalizes the "arity" field used by
the native-code compiler, in that it has room both for an arity
(always 0 in bytecode) and for the distance from the closure
to the first environment variable in the closure block.

This makes closures "self-described" and easy to scan for pointers:
everything up to the first env var is out-of-heap code pointers or
integers; everything after the first env var is a well-formed value.

At this point a bootstrap is needed.

xavierleroy committed:
Document alternate encodings for out-of-heap pointers

xavierleroy committed:
Add skip lists as a reusable data structure

The implementation is taken from globroots.c, but made reusable elsewhere.

xavierleroy committed:
Use the skip list library for global GC root management

Instead of the specialized skip list implementation that was local to
this file.

xavierleroy committed:
Use the skip list library to manage overwritten instructions in the debugger

This avoids the quadratic behaviors of the previous array-based
implementation.

Closes: #9606

gasche committed:
Changes

dra27 committed:
Add --enable-warn-error

On by default if +dev appears in VERSION and also fixed for MSVC.

dra27 committed:
Rename $gcc_warnings to $cc_warnings in configure

gasche committed:
add a testcase for 'with type <lhs> = <rhs>' scoping

gasche committed:
Typemod.merge_constraint: be careful about the typing environments

This PR fixes an old bug in the interaction between [merge_constraint]
and [Typedecl.transl_with_constraint], where
variance (and now separability) are recomputed in an invalid type
environment. See #9624 and the new tests.

lthls commented:
(There's a huge number of commits in the diff because I forgot to keep my trunk branch in sync with the upstream one, sorry about that)

Anukriti12 commented:
> (There's a huge number of commits in the diff because I forgot to keep my trunk branch in sync with the upstream one, sorry about that)

I was figuring out whether there is pass by reference but was informed later through IRC channel that there isn't anything like pointers in OCaml. Will do more changes soon

lthls commented:
Can you explain why you put `Ident.Map.empty` in some places instead of the `depth` argument from the caller ?

Apart from that, this looks like you've correctly added a new parameter to the `count` and `count_default` functions. The next step is to use this new parameter to carry the value that would otherwise be stored in `try_depth`.

I would suggest temporarily replacing the line `let rec count depth = function` with the two following lines:
```ocaml
let rec count depth lam =
  assert (depth = !try_depth);
  match lam with
```
which means that you will get an error during compilation if at some point the value for the new parameter and the old value kept in the reference disagree.

lthls commented:
(It will also force the `depth` argument to have the same type as the contents of the `try_depth` reference, which should help getting rid of the extra `Ident.Map.empty`)

Anukriti1999 commented:
Ohh this seems to be a nice approach, will add this. I added Ident.Map.empty for some debugging purpose while beginning but forgot to replace it.

gasche committed:
Merge pull request #9560 from stedolan/more-partial

More partial application warnings

gasche committed:
Typemod.merge_constraint: fix computation of the signature environment

(suggested by Leo White and Jacques Garrigue)

garrigue committed:
Fix the polymorphic recursion problem of #9603 (#9617)

nojb committed:
Fix load path ordering (#9611)

dra27 committed:
s/@raises/@raise

hhugo committed:
fix typo in #9569

garrigue committed:
Fix #7741: Failure to report escaping type variable (#9545)

garrigue committed:
add PR number for #7741

garrigue committed:
Fix #7520: Odd behaviour of refutation cases with polymorphic variants (#9547)

nojb committed:
ocamltest: do not overwrite user-defined variables

nojb committed:
Adapt tests

nojb committed:
Allow ',' at the start of OCAMLRUNPARAM

nojb committed:
Changes

xavierleroy committed:
Reword what will happen with pointers outside the heap

Following review comments.

Also: fix the assertion in the third val_of_typtr function.

gasche committed:
Merge pull request #9623 from gasche/unboxed-merge-constraint

fix an environment problem in merge_constraint

gasche committed:
Merge pull request #9629 from hhugo/patch-3

fix typo in #9569

Octachron committed:
Change entry for #9610

gasche committed:
minor Changes fix

shindere committed:
Inria CI: also test LibBFD support on FreeBSD

hannesm committed:
documentation: clarify that l=x only affects byte-code runtime

nojb committed:
Manual

gasche committed:
Merge pull request #9634 from nojb/ocamlrunparam_comma_fix

Allow ',' at the start of OCAMLRUNPARAM

nojb committed:
Changes

Octachron committed:
Merge pull request #9610 from xavierleroy/manual-intf-c

FFI documentation: naked pointers are obsolete

garrigue committed:
Fix #9640: regression introduced by #9623

garrigue committed:
update PR number

garrigue committed:
add example

xavierleroy committed:
Changes entry for #9619

garrigue committed:
factorize in merge_constraint

xavierleroy committed:
New representation of closures, intermediate bootstrap

ocamlc's computations of offsets in closures changed to match changes in the
abstract machine.

xavierleroy committed:
New representation of closures, native-code compilation

In code that builds closures, instead of the old arity field,
produce a closure information field encoding arity + position of environment.

xavierleroy committed:
Restrict 'test_locations' to 64-bit archs and update expected outputs

Expected outputs contain integer values for the "closure info"
field of some closures.  These values differ in 32 and 64 bits,
since the arity is stored in top 8 bits.  This test would need
different expected outputs for 32- and 64-bit platforms.

To keep things simple, this commit restricts the test
to only run on 64-bit platforms.  Since this changes the locations

xavierleroy committed:
During major GC, scan the environment part of closures only

Here we start reaping the benefits of the new closure representation.

The fields of a closure block that contain the code pointers need not
be scanned (in general) and must not be scanned (in no-naked-pointers mode).

Here, conservatively, we skip them in no-naked-pointers mode only,
but it would be sound to skip them unconditionally.

dra27 committed:
Fix tools/check-typo for mawk 1.3.4 (#9644)

Closes: #9643

garrigue committed:
comment for module alias + factorize with_type/with_typesubst

xavierleroy committed:
Merge pull request #9619 from xavierleroy/new-closure-repr

A self-describing representation for function closures, supporting "no naked pointers" mode and the future needs of Multicore OCaml.

xavierleroy committed:
Limit the number of parameters for an uncurried or untupled function (#9620)

This commit introduces a quantity Lambda.max_arity that is the maximal
number of parameters that a Lambda function can have.

Uncurrying is throttled so that, for example, assuming the limit is 10,
a 15-argument curried function fun x1 ... x15 -> e
becomes a 10-argument function (x1...x10) that returns a 5-argument
function (x11...x15).

Concerning untupling, a function that takes a N-tuple of arguments,
where N is above the limit, remains a function that takes a single
argument that is a tuple.

Currently, max_arity is set to 126 in native-code, to match the new
representation of closures implemented by #9619.  A signed 8-bit field
is used to store the arity.  126 instead of 127 to account for the
extra "environment" argument.

In bytecode the limit is infinity (max_int) because there are no needs
yet for a limit on the number of parameters.

gasche committed:
Merge pull request #9642 from garrigue/fix9640

Fix #9640: regression introduced by #9623

dra27 committed:
Merge pull request #9587 from gasche/Arg.Rest_list

Stdlib.Arg: a new Rest_all spec, similar to Rest, that passes all arguments at once

gasche committed:
matching: [minor] refactor the local control flow of the Unused exception

gasche committed:
Merge pull request #9608 from trefis/rematch-flattening

 pattern-matching refactoring: clarify usage of the Cannot_flatten exception

xavierleroy committed:
Add Changes for #9635

xavierleroy committed:
Upgrade the "stale" script to v3

gasche committed:
Merge pull request #9638 from hannesm/runtime-stack-limit

documentation: clarify that l=x only affects byte-code runtime

gasche committed:
Merge pull request #9514 from gasche/parmatch-exhaust-singlecol-optim-trunk

Parmatch.exhaust: single-row optimization

nojb committed:
check-typo

nojb committed:
Merge pull request #9633 from nojb/ocamltest_env_fix

ocamltest: do not overwrite user-defined variables

xavierleroy committed:
Merge pull request #9635 from xavierleroy/skiplists

Introduce a library of skip lists and use them to fix a performance issue in the debugger (issue #9606)

shindere committed:
Merge pull request #9285 from dra27/simplify-mkexe

Eliminate MKEXE_ANSI from build system

shindere committed:
Remove two oldish scripts (#9656)

shindere committed:
ocamldoc: remove debugging facility

Octachron committed:
Merge pull request #9618 from Octachron/format_documentation_precision

Document few Format quirks

gasche committed:
Merge pull request #9646 from gasche/rematch-simplify-unused

matching: refactor the local control flow of the Unused exception

gasche committed:
Merge pull request #9442 from gasche/tailcall-attribute-refactoring

[minor] refactoring the datatype for the [@tailcall] attribute

shindere committed:
Merge pull request #9661 from shindere/ocamldoc-remove-debug-support

ocamldoc: remove debugging facility

dra27 committed:
Merge pull request #9625 from dra27/warn-error

--enable-warn-error configure option

gasche committed:
Merge pull request #9602 from garrigue/document-variance-flags

add explanation for variance flags

gasche committed:
Merge pull request #9612 from garrigue/pr6744ex

fix number and add examples for #6744

jhjourdan committed:
Memprof: disable sampling when memprof is suspended. (#9628)

* Memprof: disable sampling when memprof is suspended.

* Changes.

stedolan committed:
Refactor skiplist to avoid UB casts (#9660)

This patch removes casts between struct skiplist * and struct skipcell *, and removes the "layout compatibility" fields in skiplist that were there to enable these casts.

The only algorithmic difference is that caml_skiplist_find is now "stop-at" (see discussion here), as this was slightly easier to write in the no-cast style.

shindere committed:
Remove temporary installation hacks (#9659)

* Remove temporary hack related to the Num library

This is a follow-up to commit 3de0115bfec

* Also do some cleanup in stdlib/Makefiile

xavierleroy committed:
More efficient management of code fragments (#9654)

* Introducing codefrag: a new runtime module to work with code fragments

This module collects all the operations on code fragments performed in
various places of the runtime systems.  Applies both to bytecode and
to native code.

The implementation is based on skiplists, so that "lookup fragment by
PC" and "lookup fragment by number" are efficient (logarithmic in the
number of code fragments).  "Lookup fragment by digest" remains
linear-time.

The new module also improves the handling of digests: now it is
possible to mark a code fragment as "no digest" i.e. not marshal-able.

* Use the new "codefrag" runtime module for marshaling and for the
  debugger interface

Replace the previous handling of code fragments with calls to the
functions provided by the "codefrag" runtime module.

lthls commented:
The functional part of the task is done, good job.
I'll make a number of style comments that will help making the code easier to review for other people, but the first thing we want to do is to choose a name that is closer to the original `try_depth` name. The original name meant "depth of nested try...with expressions", and the "try" part is important.
One possibility is to reuse the original `try_depth` name. Now that the original reference has been removed, there is no ambiguity left, and it will make it very clear that it's used for the same purpose as the old reference.
I'll make a number of small inline comments about comments, white space, and stuff like that too.

lthls commented on lambda/simplif.ml:
```diff
@@ -102,8 +102,8 @@ let simplify_exits lam =
-
-  let try_depth = ref 0 in
+(* 
+  let try_depth = ref 0 in *)
```
Comment should be removed for the final version.

lthls commented on lambda/simplif.ml:
```diff
@@ -119,69 +119,76 @@ let simplify_exits lam =
+  let rec count depth = function
+  (* let rec count depth lam =
+    assert (depth = !try_depth);
+    match lam with *)
```
Comment should be removed for the final version.

lthls commented on lambda/simplif.ml:
```diff
@@ -119,69 +119,76 @@ let simplify_exits lam =
   | (Lvar _| Lconst _) -> ()
-  | Lapply ap -> count ap.ap_func; List.iter count ap.ap_args
-  | Lfunction {body} -> count body
+  | Lapply ap -> (count depth) ap.ap_func; List.iter (count depth) ap.ap_args
```
The first pair of parentheses is not needed.

lthls commented on lambda/simplif.ml:
```diff
@@ -119,69 +119,76 @@ let simplify_exits lam =
+        | _ -> count depth d; count depth d (* default will get replicated *)
       end
-  | Lstaticraise (i,ls) -> incr_exit i 1 !try_depth; List.iter count ls
+  
```
There's an extra empty line here, that you should remove for consistency.

lthls commented on lambda/simplif.ml:
```diff
@@ -119,69 +119,76 @@ let simplify_exits lam =
   in
-  count lam;
-  assert(!try_depth = 0);
+
```
Extra empty line, might be better to remove it.

lthls commented on lambda/simplif.ml:
```diff
@@ -201,19 +208,19 @@ let simplify_exits lam =
   let subst = Hashtbl.create 17 in
-
-  let rec simplif = function
+  
```
There are two extra spaces on this line, most likely added by your editor; you should remove them (and keep the empty line).

lthls commented on lambda/simplif.ml:
```diff
@@ -201,19 +208,19 @@ let simplify_exits lam =
   | Lapply ap ->
-      Lapply{ap with ap_func = simplif ap.ap_func;
-                     ap_args = List.map simplif ap.ap_args}
+      Lapply{ap with ap_func = (simplif depth) ap.ap_func;
```
These parentheses are not needed, I would suggest removing them.

lthls commented on lambda/simplif.ml:
```diff
@@ -294,38 +301,38 @@ let simplify_exits lam =
-  | Lsequence(l1, l2) -> Lsequence(simplif l1, simplif l2)
-  | Lwhile(l1, l2) -> Lwhile(simplif l1, simplif l2)
+      let depth1 = depth +1 in
+      let l1 = simplif depth1 l1 in
```
Using `simplif (depth + 1) l1` and removing the `depth1` let-binding above probably better here.

lthls reviewed (commented):
I've pointed out all the places where I think some cleanup should be done. I think once you've done that we should think of submitting back the PR to the main repository, but that's not urgent.

lthls commented:
I think we can now make a pull request on the main repository with your changes.
There's no guarantee that it will be merged, because it doesn't fix any bug or add any feature, but it could give you an idea of what kind of feedback to expect.
My own advice would be to submit a pull request but not spend too much time on it (in particular, you don't have to reply to every comment immediately), but I'll leave the choice to you.
I could also submit the PR on your behalf, if you prefer it this way.

Anukriti1999 commented:
No, that's not a concern for me. I will be making the first PR today itself.

Anukriti12 committed:
Removed last commit

Anukriti12 committed:
made try_depth a labeled argument

Anukriti12 committed:
tried to remove simplif files and added updatd version of linearize.ml

Anukriti12 committed:
removed previous_exit_labels and updated with changes
